core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java

```diff
@@ -613,6 +613,59 @@ public static SqlCall stripSeparator(SqlCall call) {
   public static final SqlReturnTypeInference ARG0_EXCEPT_INTEGER_NULLABLE =
       ARG0_EXCEPT_INTEGER.andThen(SqlTypeTransforms.TO_NULLABLE);
 
+  /**
+   * Returns the type of the first non NULL argument or INTEGER if all arguments
+   * are NULL.
+   */
+  public static final SqlReturnTypeInference LARGEST_INT_OR_FIRST_NON_NULL_DEFAULT_INTEGER =
+      opBinding -> {
+        final RelDataTypeFactory typeFactory = opBinding.getTypeFactory();
+        RelDataType integerReturnType = null;
+        boolean allArgsInteger = true;
+        for (RelDataType opType : opBinding.collectOperandTypes()) {
+          if (SqlTypeName.INT_TYPES.contains(opType.getSqlTypeName())) {
+            if (integerReturnType == null) {
+              integerReturnType = opType;
+            } else if (isIntegerTypeLarger(integerReturnType, opType)) {
+              integerReturnType = opType;
+            }
+          } else {
+            allArgsInteger = false;
+            break;
+          }
+        }
+        if (integerReturnType != null && allArgsInteger) {
+          return typeFactory.createTypeWithNullability(integerReturnType, true);
```

Comment:
why are all these types nullable?

Reply:
Since `NULL` is a possible result.
```
BITAND(1, NULL) = NULL
BITAND(CAST(x'01' AS BINARY(1)), NULL) = NULL
```

If nullability is set to `false` here, then `NULL` results are coerced back to integer types with the value `0`.

Reply:
I see what you mean. I have reworked the return type detection. It is now only nullable if at least one argument type is nullable.